File: c/context.c
#include <stdio.h>
#include <time.h>
#include <sys/timeb.h>
#include <unistd.h>

#include "context.h"

FILE *_context_log;

void ctx_set_log(const char *name, int append) {
	if(name) {
		_context_log = fopen(name, append ? "a" : "w");
	}
	else {
		if(_context_log) {
			fclose(_context_log);
		}
		_context_log = NULL;
	}
}

void ctx_log_msg(const char *func, const char *text, const char *type) {
	if(_context_log) {
		struct timeb tmb;
		char hostname[256];

		ftime(&tmb);
		gethostname(hostname, 256);

        fprintf(
			_context_log,
			"%ld.%d %s %d %d %s %s %s\n",
            tmb.time, tmb.millitm,
            hostname,
			getpid(),
			getpid(), //gettid(),
            type, func, text
        );
	}
}

File: php/context.php
<?php
$_context_log = null;

function ctx_set_log($url, $append=True) {
	global $_context_log;
	if($url) {
		$p = parse_url($url);
		if($p["scheme"] == "file") {
			if(@$p["host"] && @$p["path"]) {
				$path = $p["host"] . $p["path"];
			}
			elseif(@$p["host"]) {
				$path = $p["host"];
			}
			elseif(@$p["path"]) {
				$path = $p["path"];
			}
			else {
				throw new Exception("Filename not recognised");
			}
			$_context_log = fopen($path, $append ? "a" : "w");
		}
		else {
			throw new Exception("Only file:// protocol is supported");
		}
	}
	else {
		$_context_log = null;
	}
}

function ctx_log_msg($func, $text, $type) {
	global $_context_log;
	if($_context_log) {
        fprintf(
			$_context_log,
			"%f %s %d %d %s %s %s\n",
            microtime(true), # returning a float is 5.0+
            php_uname('n'),  # gethostname() is 5.3+
			posix_getpid(),
			function_exists("hphp_get_thread_id") ? hphp_get_thread_id() : posix_getpid(),
            $type, $func, $text
        );
	}
}

function __get_func() {
	$stack = debug_backtrace();
	if(count($stack) < 3) {
		return "top-level";
	}
	$p = $stack[2];
	return $p['function'];
}

function ctx_log_bmark($text=null) {ctx_log_msg(__get_func(), $text, "BMARK");}
function ctx_log_clear($text=null) {ctx_log_msg(__get_func(), $text, "CLEAR");}
function ctx_log_endok($text=null) {ctx_log_msg(__get_func(), $text, "ENDOK");}
function ctx_log_ender($text=null) {ctx_log_msg(__get_func(), $text, "ENDER");}

function ctx_log_start($text=null, $bookmark=false, $clear=false) {
	if($clear) {
		ctx_log_msg(__get_func(), $text, "CLEAR");
	}
	if($bookmark) {
		ctx_log_msg(__get_func(), $text, "BMARK");
	}
	ctx_log_msg(__get_func(), $text, "START");
}
?>
